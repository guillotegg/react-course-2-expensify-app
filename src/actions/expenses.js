import uuid from 'uuid';

export const addExpense = ({ id = uuid(),
                             description = '', 
                             note = '', 
                             amount = 0, 
                             createdAt = 0 } = {} 
                          ) => ({
    'type' : 'ADD_EXPENSE',
    'expense' : { id, description, note, amount, createdAt}
});

export const removeExpense = ( id = 0 ) => ({
    'type': 'REMOVE_EXPENSE',
    'id' : id
});

export const editExpense = ( id = 0, updates ) => ({
    'type': 'EDIT_EXPENSE',
    id,
    updates
});

