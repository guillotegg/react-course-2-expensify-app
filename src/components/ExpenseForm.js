import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import { timingSafeEqual } from 'crypto';
import { throws } from 'assert';

const now = moment();

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);       
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    };

    OnDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    OnNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d*((\.|,)\d{0,2})?$/ )) {
            this.setState(() => ({ amount }));
        }        
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({focused}) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount){
            this.setState(() => ({ error: 'Please provide description and amount'}));
        } else {
            this.setState(() => ( { error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf()
            });
        }        
    };
    render() {
        return (
            <div>
            {this.state.error && <div>{this.state.error}</div>}
            <form onSubmit={this.onSubmit}>
                    <input 
                        type="text" 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.OnDescriptionChange} 
                    />
                    <input 
                        type="number" 
                        placeholder="Amount" 
                        value={this.state.amount} 
                        onChange={this.onAmountChange} 
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt} 
                        onDateChange={this.onDateChange} 
                        focused={this.state.calendarFocused} 
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />                    
                    <textarea 
                        placeholder="Add a note for your expense" 
                        value={this.state.note} 
                        onChange={this.OnNoteChange} 
                    />
                    <button>Add Expense</button> 
                </form>
            </div>
        )
    }
}