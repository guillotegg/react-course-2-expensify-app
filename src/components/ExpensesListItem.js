import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const ExpensesListItem = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <div>{amount} - {createdAt}</div>
    </div>
);

export default connect()(ExpensesListItem);