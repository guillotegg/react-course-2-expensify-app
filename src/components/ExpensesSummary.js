import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expenseTotal}) => (
    <div>
        <h1>Viewing {expenseCount} expenses totalling {numeral(expenseTotal / 100).format('$0,0.00')}</h1>
    </div>
);

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: expenses.length,
        expenseTotal: getExpensesTotal(expenses)
    };
}

export default connect(mapStateToProps)(ExpensesSummary);