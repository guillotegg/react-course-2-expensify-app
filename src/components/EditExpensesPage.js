import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses.js';

export class EditExpensesPage extends React.Component {
    onSubmit = (expense) => {              
        this.props.editExpense(this.props.expense.expenseId, expense);
        this.props.history.push('/');
    };
    onClick = (id) => {
        this.props.removeExpense(id);
        this.props.history.push('/');
    };
    render(expense) {
        return (
            <div>
               <ExpenseForm expense={expense} onSubmit={this.onSubmit} />
               <button onClick={this.onClick}>Remove</button> 
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensesPage);
