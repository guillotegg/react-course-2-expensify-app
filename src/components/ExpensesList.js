import React from 'react';
import { connect } from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import selectExpenses from './../selectors/expenses';

export const ExpensesList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpensesListItem {...expense} />
                })
            )
        }
        
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpensesList);