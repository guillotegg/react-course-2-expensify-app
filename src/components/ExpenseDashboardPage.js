import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesListFilters from './ExpensesListFilters';

export const ExpenseDashboardPage = () => (
    <div>
        <ExpensesListFilters />    
        <ExpensesList />
    </div>
);

export default ExpenseDashboardPage;