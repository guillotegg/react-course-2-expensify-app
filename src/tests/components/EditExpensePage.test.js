import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensesPage } from '../../components/EditExpensesPage';
import expenses from '../fixtures/expenses';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensesPage expense={expenses[0]} editExpense={editExpense} removeExpense={removeExpense} history={history} />);
});


test('should render EditExpensesPage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {   
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].expenseId, expenses[1]);
});


test('should handle removeExpense', () => {
    const id = expenses[1].expenseId;
    wrapper.find('button').prop('onClick')(id);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith(id);
});


