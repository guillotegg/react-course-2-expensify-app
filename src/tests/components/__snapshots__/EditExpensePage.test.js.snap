// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensesPage 1`] = `
<div>
  <ExpenseForm
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
